File: src/boj/boj_9663.java
package boj;

import java.io.BufferedReader;
import java.io.InputStreamReader;

public class boj_9663 {
    static int answer = 0;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        backTracking(n, 0, new int[n]);
        System.out.println(answer);
        br.close();
    }

    static void backTracking(int n, int r, int[] col) {
        if (!isPossible(r - 1, col)) {
            return;
        } else if (n == r) {
            answer++;
            return;
        }

        for (int i = 0; i < n; i++) {
            col[r] = i;
            backTracking(n, r + 1, col);
        }
    }

    static boolean isPossible(int r, int[] col) {
        for (int i = 0; i < r; i++) {
            boolean isSameCol = col[i] == col[r];
            boolean isDiagonal = Math.abs(col[i] - col[r]) == Math.abs(i - r);
            if (isSameCol || isDiagonal) {
                return false;
            }
        }
        return true;
    }
}

File: src/boj/boj_9663.kt
package boj

import java.io.BufferedReader
import java.io.InputStreamReader

private lateinit var array: IntArray
private var n = 0
private var answer = 0

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    n = readLine().toInt()
    array = IntArray(n)
    bt(0)
    println(answer)
    close()
}

private fun bt(depth: Int) {
    if (depth == n) {
        answer++
        return
    }

    for (i in 0 until n) {
        val idx = depth * n + i
        if (possible(depth, idx) && (depth == 0 || array[depth - 1] < idx)) {
            array[depth] = idx
            bt(depth + 1)
        }
    }
}

private fun possible(depth: Int, idx: Int): Boolean {
    if (depth == 0) {
        return true
    }
    for (i in 0 until depth) {
        val diffRow = idx / n - array[i] / n
        val diffCol = idx % n - array[i] % n
        when {
            diffRow == 0 -> return false
            diffCol == 0 -> return false
            diffRow - diffCol == 0 -> return false
            diffRow + diffCol == 0 -> return false
        }
    }
    return true
}